test/unit/org/apache/cassandra/index/sai/cql/IndexCompressionTest.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.cassandra.index.sai.SAITester;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexComponentType;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.io.compress.CompressionMetadata;
+import org.apache.cassandra.io.compress.ZstdCompressor;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.schema.IndexMetadata;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class IndexCompressionTest extends SAITester
+{
+    @BeforeClass
+    public static void setup()
+    {
+        CassandraRelevantProperties.INDEX_COMPRESSION_ENABLED.setBoolean(true);
+    }
+
+    @Test
+    public void testKeyCompression()
+    {
+        createTable("CREATE TABLE %s (pk int, c text, val text, PRIMARY KEY(pk, c))");
+        String indexName = createIndex("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex' WITH key_compression = {'class': 'LZ4Compressor'}");
+        for (int i = 0; i < 1000; i++)
+            execute("INSERT INTO %s(pk, c, val) VALUES (?, ?, ?)", i, "key", "value" + i);
+
+        flush();
+
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value0'"), 1);
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value5'"), 1);
+        assertRowCount(execute("SELECT * FROM %s WHERE val = 'value999'"), 1);
+
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        StorageAttachedIndex sai = (StorageAttachedIndex) cfs.indexManager.getIndexByName(indexName);
+        IndexContext context = sai.getIndexContext();
+
+        for (SSTableReader sstable : cfs.getLiveSSTables())
+        {
+            IndexDescriptor descriptor = IndexDescriptor.load(sstable, Set.of(context));
+            assertCompressed(descriptor.perSSTableComponents().get(IndexComponentType.PRIMARY_KEY_TRIE));
+            assertCompressed(descriptor.perSSTableComponents().get(IndexComponentType.PRIMARY_KEY_BLOCKS));
+        }
+    }
+
+    @Test
+    public void testKeyCompressionIsCommonForAllIndexes()
+    {
+        // When creating another index with different key compression, the key compression of the first index
+        // gets overwritten by the compression settings of the second index.
+        // This is beacuse both indexes share the same primary key map, and it can be compressed in one way only.
+        createTable("CREATE TABLE %s (pk int, c text, val1 text, val2 text, PRIMARY KEY(pk, c))");
+        String index1 = createIndex("CREATE CUSTOM INDEX ON %s(val1) USING 'StorageAttachedIndex' WITH key_compression = {'class': 'LZ4Compressor'}");
+        String index2 = createIndex("CREATE CUSTOM INDEX ON %s(val2) USING 'StorageAttachedIndex' WITH key_compression = {'class': 'ZstdCompressor'}");
```

Comment:
We can quickly check here that the client warning is emitted:
```suggestion
        ClientWarn.instance.captureWarnings();
        String index1 = createIndex("CREATE CUSTOM INDEX ON %s(val1) USING 'StorageAttachedIndex' WITH key_compression = {'class': 'LZ4Compressor'}");
        Assertions.assertThat(ClientWarn.instance.getWarnings()).isNull();

        String index2 = createIndex("CREATE CUSTOM INDEX ON %s(val2) USING 'StorageAttachedIndex' WITH key_compression = {'class': 'ZstdCompressor'}");
        Assertions.assertThat(ClientWarn.instance.getWarnings()).hasSize(1).allMatch(e -> e.contains(index1));
```